External Media: Improve Insert Flow

obenland opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->
Improves the insert flow by not taking the user out of the media modal while inserting.

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Removes Copy/Insert Modal
* Updates Insert button to say "Select" (like in media library) and be busy while inserting.
* Updates media items to show spinner while they're being copied.

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Check out this PR and run `yarn run build-extensions`
* In the Editor, insert a media block (Cover, Media & Text, Image, Gallery)
* Click "Select [Image(s)|Media]" and chose either Pexels or Google Photos.
* Based on the block, select one or more images and click "Select".
* Observe the spinner overlay over the thumbnail images and the Select button now being busy and saying "Inserting…".

**After**
![image](https://user-images.githubusercontent.com/1398304/83928188-b46e5b00-a743-11ea-9f2a-afcaa0387c10.png)


#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
Not needed.

Closes #15867.
Fixes #15867.

obenland committed:
[not verified] # This is a combination of 5 commits.

External Media: Improve Insert Flow

See https://github.com/Automattic/jetpack/issues/15867

Remove unused import

Lock pointer events when copying
This prevents a UX issue where user can still interact with the modal content while the "Inserting..." request is pending.

Announce Inserting action along with alt text of selected images.

Update modal headers to reflect status

Props @jancavan.

WunderBart reviewed (commented):
Nice UX improvement and code simplification at the same time, thanks!

I've noticed that I'm able to interact with the modal content while the copy request is pending so I decided to add a small improvement for that in https://github.com/Automattic/jetpack/pull/16081/commits/18db7348a5ed5d43d1f5410f99665af03cf44aca. I'd approve this PR but since I pushed myself it would be good for someone else to take a look.

jeryj commented:
I agree with Bart - this feels a lot nicer than the Copying Modal flow 🙌

In https://github.com/Automattic/jetpack/pull/16081/commits/38fa4a9e4c7022426b43448ce563319e8db2a2f3, I added a `speak` announcement on the copying action to announce "Inserting: %s" where %s is a `join(', ')` string of all images' alt text.

jeryj commented on extensions/shared/external-media/editor.scss:
```diff
@@ -21,39 +21,6 @@ $grid-size: 8px;
-	}
-}
-
 .jetpack-external-media-browser:not( .jetpack-external-media-browser__is-copying ) {
```
I wonder if we can remove this entire section, or at minimum, the `:not(...)` selector.

jancavan commented:
Thanks @obenland! There's a couple more things that I think we should add to better the experience. When images are being copied:

- Thumbnails with the spinner should say "Inserting image..." to make it more obvious what's happening here in case the user misses the change in button state:

![Screen Shot 2020-06-08 at 1 57 04 PM](https://user-images.githubusercontent.com/4924246/84079759-f25fbf00-a98f-11ea-9a2b-b206c289e577.png)

- Search field and thumbanils should be disabled. The only thing the user should be able to do here is dismiss the modal (as an escape hatch should the copying take a long time):

![Screen Shot 2020-06-08 at 1 55 45 PM](https://user-images.githubusercontent.com/4924246/84079638-c0e6f380-a98f-11ea-8ca7-7e9885790cea.png)

## Prototype

![ezgif-6-060e0960e461](https://user-images.githubusercontent.com/4924246/84080514-21c2fb80-a991-11ea-92d7-d0f61739ee99.gif)

Here's a clickable prototype to get a feel for it. You can select the **2nd** and **4th** thumbnail:

https://www.figma.com/proto/ERNU68IlQZwaaDazIGfNL7/Ajax-Improve-External-Media-Flow?node-id=52%3A42&viewport=-6877%2C-4747%2C0.5&scaling=min-zoom

jancavan commented:
- Modal heading also needs to change to "Insert media" once images are being copied over:

![Screen Shot 2020-06-08 at 3 27 23 PM](https://user-images.githubusercontent.com/4924246/84086359-90598680-a99c-11ea-951d-90e720e8a6df.png)

- Should be "Select media" with lowecase `m`:

![Screen Shot 2020-06-08 at 3 43 44 PM](https://user-images.githubusercontent.com/4924246/84087386-d879a880-a99e-11ea-81b5-afa08fac3e99.png)

marekhrabe commented on extensions/shared/external-media/editor.scss:
```diff
@@ -71,32 +38,17 @@ $grid-size: 8px;
 	}
 }
 
+.jetpack-external-media-browser__is-copying {
+	pointer-events: none;
+}
+
```
While this is ok, everything will still be accessible and functional through keyboard. I think we should set our interactive elements as disabled, based on the same flag.

marekhrabe reviewed (approved):
Addressed:
- [x] disabled states for:  #16103
  - search input and button
  - thumbnails
  - pagination button
  - filters in google photos
- [x] append "Inserting image…" text to the loading state of thumbnail #16103

Todos from the feedback:
- [ ] add number of selected images next to the insert button (only show when multi selections are enabled)
- [ ] see if we can clean and simplify our css, as we now have much less changes for the "copying" state

I wouldn't mind if those are addressed separately, to keep the flow of PRs moving. Otherwise, this PR looks good to me and works well.

jancavan commented:
@marekhrabe Would also like to add this [copy change](https://github.com/Automattic/jetpack/pull/16081#issuecomment-640925924) in, if possible. Thanks.

obenland commented:
> Would also like to add this copy change in, if possible. Thanks.

Thank you! Fixed in 96ee699.

jeryj committed:
External Media: Fix mobile modal size (#16098)

* External Media: Improve Insert Flow

See https://github.com/Automattic/jetpack/issues/15867

* Remove unused import

* Lock pointer events when copying
This prevents a UX issue where user can still interact with the modal content while the "Inserting..." request is pending.

* Announce Inserting action along with alt text of selected images.

* Fix mobile mobile width to take up the full available modal space

* Change is-copying to modifier instead of element to follow BEM syntax

Co-authored-by: Konstantin Obenland <obenland@gmx.de>
Co-authored-by: Bart <bartlomiej.kalisz@gmail.com>

obenland committed:
Merge branch 'feature/external-media' into update/insert-flow

marekhrabe committed:
External Media: Loading and disabled states for the copying state (#16103)

* disable pagination while loading

* fade out all images when copying

* disable event handlers and mark as disabled when copying

* disable pexels search when copying

* disabled google filters when copying

* add text to copying indicator

* disable google photos view selector when copying

* [not verified] remove debugging condition

* Pass functions not booleans to event emitters

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

obenland commented:
I noticed the modal scrolling back to top when inserting an image from below the fold. With the dedicated upload modal that was probably desired—now, not anymore. 

@marekhrabe Does that sound familiar? I've tried tracking down what might cause it but I've not come very far.

jeryj committed:
Removed :not() copying css state since they are the same modal now

jeryj committed:
Reset focus to the modal after clicking Insert. Otherwise focus is reset to the body and focus needs to stay trapped within the modal.

obenland committed:
Cosmetically rearranging code order

jeryj committed:
External Media: Screen Reader accessible states for Insert flow (#16113)

* Add visually hidden description for the external media modal to give improved screen reader instructions and communicate the copying state

* [not verified] Remove title attribute from grid images

* [not verified] Move Inserting Image... copy to before the image so it is announced first.

* Separate-out description and element id

Helps with keeping describedby id in sync

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

jeryj reviewed (approved):
:shipit:
